juju/juju
# [JUJU-528] Allow the juju snap to be strictly confined

This PR allows the juju snap to be built as strictly confined. It is still classic by default, but the `snapcraft.yaml` can be changed to be strict and the resulting snap will function as expected.

New interfaces were needed in snapcraft. Most are not auto connected; we need to ask for exceptions in the snapcraft forum. Until then, they will need to be connected manually as per the QA steps.

The other main change is to tweak how the user home dir is located. We need to use `SNAP_REAL_HOME`. This has been done upstream in `juju/utils/v3`.

There's also a couple of drive by fixes to correctly use `juju/utils/v3`.

TODO
In strict mode, microk8s is not supported because juju needs to get the microk8s config and status and strictly confined snaps cannot run external commands. We need extra work to get the juju and microk8s snaps talking. For now, microk8s is disabled if the snap is strictly confined.

TODO
even though it's grant read access via `personal-files`, there's a permission issue reading one of the possible lxc config files (the other similarly configured one works fine)
```
WARNING unable to read/parse LXC config file (/home/ian/snap/lxd/current/.config/lxc/config.yml): open /home/ian/snap/lxd/current/.config/lxc/config.yml: permission denied
```

## QA steps

I used 2.9.24 below but change to whatever version is built.
Edit `snapcraft.yaml` and set `confinement: strict`
```
snapcraft remote-build --launchpad-accept-public-upload --build-on amd64
sudo snap install ./juju_2.9.24_amd64.snap --dangerous
sudo snap connect juju:lxd lxd:lxd
sudo snap connect juju:ssh-keys
sudo snap connect juju:config-juju
sudo snap connect juju:config-lxd
sudo snap connect juju:cloud-credentials-juju
```
clouds and autoload-credentials should read existing info in ~/.local/share/juju and ~/.aws or ~/.novarc etc and present the same choices as when a normal juju binary is run
```
/snap/bin/juju clouds
/snap/bin/juju autoload-credentials
```
add cloud to the client
create a cloud yaml file somewhere in home directory
```
juju add-cloud --client mycloud /path/to/cloud.yaml
```
after bootstrap the ssh-keys command should list your normal ssh key as being uploaded
```
/snap/bin/juju bootstrap lxd test
/snap/bin/juju ssh-keys
/snap/bin/juju deploy ubuntu
/snap/bin/juju create-backup
/snap/bin/kill-controller -y -t 0 test
```
